Fix activation for nested conductor / screen scenarios

Yinimi opened the pull request:
There have been some discussions and changes related to when OnActivate() is called.
#593 He wanted to use IsInitialized / IsActive to control the visibility of some loading indicator. If the property is set to true before the OnActivate / OnInitialize call it will obviously not be able to use it for determining if the loading has finished.
I can understand the reason for this request but I think a method called OnActivate() should be called after the IsActive is set to true, otherwise Activating or BeforeActivate are better names in my opinion.

Since some devs (like me) want to use OnActivate() to trigger the initializing of the ViewModel there are scenarios where it is problematic if the IsActive is not set yet. I created the [ActivateWhileActivateOneLevel()](https://github.com/Yinimi/Caliburn.Micro/blob/e246f672cd93b0340f3792eefdfa58d955c6a18e/src/Caliburn.Micro.Core.Tests/ConductorWithCollectionOneActiveTests.cs#L125) test method to demonstrate it. 
To provide such a method @erik-hooper added with #859 a [OnActivate**d**()](https://github.com/Caliburn-Micro/Caliburn.Micro/pull/859/commits/5a524b0904ac59dcb43425e286764a4e42619a7c) method which is called after the property has been set.
If I use this method with another test [ActivatedWhileActivatedOneLevel()](https://github.com/Yinimi/Caliburn.Micro/blob/e246f672cd93b0340f3792eefdfa58d955c6a18e/src/Caliburn.Micro.Core.Tests/ConductorWithCollectionOneActiveTests.cs#L148C27-L148C58) The issue is resolved.

I have a more complicated and nested scenario. [example Test](https://github.com/Yinimi/Caliburn.Micro/blob/e246f672cd93b0340f3792eefdfa58d955c6a18e/src/Caliburn.Micro.Core.Tests/ConductorWithCollectionOneActiveTests.cs#L171)
* Outer Conductor
  * some Page
  * Inner Conductor
    * View 1
    * View 2

In my app the inner conductor handles the navigation depending on some configuration.
The outer conductor handles basic stuff and also the display of the user preferences ("some Page").
If the View 1 now wants automatically to continue after the user preferences have been shown (think of auto login was just enabled) then in this nested scenario the View 2 will not be initialized because all this is triggered by the OnActivate() method of the inner conductor.

With this change the behavior should be like before the changes of #593 where the active item(s) are activated after the conductor has IsActive enabled.

In #789 it was also discussed to add something like BeforeActivate and AfterActivate.
I added a [OnInitializedAsync()](https://github.com/Yinimi/Caliburn.Micro/blob/b0cbfc64eca3da3e94d32db6149e816bf05089c1/src/Caliburn.Micro.Core/Screen.cs#L107) to adress this as well. I didn't touch the OnViewLoaded (it was wished to make this async), this looks more complicated.

vb2ae commented:
Yes the release is version 5

vb2ae commented on src/Caliburn.Micro.Core/ConductorWithCollectionOneActive.cs:
I like the OnActivated name better but could you leave OnActivate and mark it as obsolete? This will keep the amount of code we break when someone updates the caliburn micro version in their project to a minimum

Yinimi commented:
I added the obsolete messages

Yinimi commented:
I think #764 Is also resolved with this change

a44281071 commented:
asp.net Blazor use ```void OnAfterRender()``` and ```Task OnAfterRenderAsync()``` for view update. And use ```void OnInitialized()``` and ```Task OnInitializedAsync()``` for init load.

ASP.NET Core Razor component lifecycle:
> [Component initialization](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0#component-initialization-oninitializedasync)

Yinimi commented:
If I read the build pipeline log correct the android 28 sdk is missing? Is the pipeline running on a specific machine with a build agent? Then it would probably be required to install it there manually

vb2ae commented:
Working on a fix for android version in another branch

vb2ae commented on src/Caliburn.Micro.Core.Tests/ConductorWithCollectionOneActiveTests.cs:
```diff
@@ -24,6 +24,46 @@ protected override async Task OnDeactivateAsync(bool close, CancellationToken ca
+        {
+            public event EventHandler RequestContinue;
+
+            public int ActivateCalledCount { get; private set; }
```
Is this just to see show the user how many times Activate is called?

vb2ae commented on src/Caliburn.Micro.Core/Screen.cs:
```diff
@@ -173,14 +174,24 @@ public virtual async Task TryCloseAsync(bool? dialogResult = null)
         /// <summary>
         /// Called when initializing.
         /// </summary>
+        [Obsolete("Override OnInitializedAsync")]
```
Thanks for marking old method and marking obsolete

vb2ae reviewed (changes requested):
Looks good.  Could you please update the Feature apps with the change

Yinimi commented:
@vb2ae I updated the Feature branch.
Additionally I fixed the scenarios solution but it looked very outdated, it was still using some caliburn micro version 3 dlls.

vb2ae commented:
@Yinimi thank you the samples do need help.

Yinimi commented:
There seems to be a related issue #906 
The changes cause unit tests to fail

vb2ae commented:
@Yinimi I will look at it

vb2ae commented on src/Caliburn.Micro.Core/ConductorBaseWithActiveItem.cs:
```diff
@@ -39,12 +39,13 @@ object IHaveActiveItem.ActiveItem
         /// <returns>A task that represents the asynchronous operation.</returns>
         protected virtual async Task ChangeActiveItemAsync(T newItem, bool closePrevious, CancellationToken cancellationToken)
         {
+            var previousItem = _activeItem;
```
Good find

vb2ae reviewed (approved):
Thank you for finding this

a44281071 commented:
Maybe a 'state' like connection is better. Just whether one is activated or not expresses too little meaning.
Pending. InProgress. Completed, Failed...

For example, "In Progress" indicates that it is activating. 

For CM.v5, the func name 'activating' and 'Activated/AfterActivated' is better. We don't need 'Active' means it is being activated.
“Don't let the doubt persist”. Just tell us clearly in the tutorial documentation and let us replace it globally.

vb2ae commented:
@a44281071 could you create an issue for the requested change?
